lib/finch/http1/conn.ex

```diff
@@ -183,23 +189,25 @@ defmodule Finch.Conn do
       {:error, mint, error, {status, headers}} ->
         metadata = Map.merge(metadata, %{error: error, status: status, headers: headers})
         Telemetry.stop(:recv, start_time, metadata, extra_measurements)
-        {:error, %{conn | mint: mint}, error}
+        {:error, close(%{conn | mint: mint}), error}
     end
   end
 
-  defp receive_response(entries, acc, fun, mint, ref, timeout, status \\ nil, headers \\ [])
+  defp receive_response(entries, acc, fun, mint, ref, timeout, max_body, status \\ nil, headers \\ [], body_sum \\ 0)
 
-  defp receive_response([], acc, fun, mint, ref, timeout, status, headers) do
-    case MintHTTP1.recv(mint, 0, timeout) do
+  defp receive_response([], acc, fun, mint, ref, timeout, max_body, status, headers, body_sum) do
+    num_bytes_to_receive = if max_body === :infinity, do: 0, else: min(@default_tcp_recbuf_size, max_body - body_sum)
```

Comment:
@sneako I'll run a simple experiment with `strace` in the next hour to determine what happens at the system call level if I set this to `max_body - body_sum` instead of calculating the minimum.